homeassistant/components/matter/media_player.py

```diff
@@ -0,0 +1,137 @@
+
+    entity_description: MatterMediaPlayerEntityDescription
+    _platform_translation_key = "media_player"
+    _attr_supported_features = (
+        MediaPlayerEntityFeature.VOLUME_SET
+        | MediaPlayerEntityFeature.VOLUME_MUTE
+        | MediaPlayerEntityFeature.VOLUME_STEP
+    )
+    _attr_volume_step = 0.01
+    # No state in the Speaker endpoint as it is dedicated to volume control
+    _attr_state = MediaPlayerState.ON
+
+    async def async_set_volume_level(self, volume: float) -> None:
+        """Set volume level."""
+        # Read device max level (defaults to 254 if unavailable)
+        max_level = self.get_matter_attribute_value(
+            clusters.LevelControl.Attributes.MaxLevel
+        )
+        if max_level is None or max_level <= 0:
+            max_level = 254
+
+        if volume == 0:
+            # If volume level is 0, turn off the volume
+            await self.async_mute_volume(True)
+            return
+        # Unmute the volume if it was muted
+        await self.async_mute_volume(False)
+
+        # Convert HA volume level (0-1) to Matter level (0-254)
+        # Matter uses 0-254 for volume levels, where 0 is off
+        matter_volume_level = int(volume * max_level)
+        await self.send_device_command(
+            clusters.LevelControl.Commands.MoveToLevel(level=matter_volume_level)
+        )
+
+    async def async_mute_volume(self, mute: bool) -> None:
+        """Mute or unmute the media player."""
+        # OnOff attribute == True state means volume is on, so HA should show mute switch as off
+        if mute:
+            # Send Matter Off command
+            await self.send_device_command(
+                clusters.OnOff.Commands.Off(),
+            )
+        # OnOff attribute == False means volume is off (muted), so HA should show mute switch as on
+        else:
+            await self.send_device_command(
+                clusters.OnOff.Commands.On(),
+            )
+
+    @callback
+    def _update_from_device(self) -> None:
+        """Update from device."""
+        matter_volume = self.get_matter_attribute_value(
+            clusters.LevelControl.Attributes.CurrentLevel
+        )
+        LOGGER.debug(
+            "matter_volume: %s for %s",
+            matter_volume,
+            self.entity_id,
+        )
+        if matter_volume is None:
+            # Attribute unavailable, cannot determine volume state
+            return
+        if matter_volume == 0:
+            self._attr_is_volume_muted = True
+        # Convert Matter CurrentLevel (0-254) to HA volume level (0-1)
+        else:
+            self._attr_is_volume_muted = False
+            max_level = self.get_matter_attribute_value(
+                clusters.LevelControl.Attributes.MaxLevel
+            )
+            if max_level is None or max_level <= 0:
+                max_level = 254
+            self._attr_volume_level = matter_volume / float(max_level)
```

Comment:
Related to the above comment: We only determine if we're muted or not depending on the `CurrentLevel` attribute. We never actually look at the `OnOff.Attributes.OnOff` state. 

If this works as expected, I think it's all the more reason to remove the `OnOff.Commands.On()` call above, as we already rely on the `OnOff`/`LevelControl` linking behavior of the device here.

Reply:
Actually, if a speaker gets muted via another Matter controller or using device buttons, do we know it sends an attribute report with `CurrentLevel`? Perhaps it only sends an `OnOff` attribute report. We would ignore the latter completely.